File: golang-backend/internal/service/commit_service.go
package service

import (
	"fmt"
	"log/slog"
	"sort"
	"strings"
	"time"

	"github.com/huangjin/coding-history/golang-backend/internal/dto"
	"github.com/huangjin/coding-history/golang-backend/internal/githelper"
	"github.com/huangjin/coding-history/golang-backend/internal/model"
	"github.com/huangjin/coding-history/golang-backend/internal/workstatus"
	"gorm.io/gorm"
)

type CommitService struct {
	db                 *gorm.DB
	dataMetricsConfigSvc *DataMetricsConfigService
}

func NewCommitService(db *gorm.DB) *CommitService {
	return &CommitService{
		db:                 db,
		dataMetricsConfigSvc: NewDataMetricsConfigService(db),
	}
}

func (s *CommitService) GetMaxCommitDateMsForRepo(repoID string) (int64, error) {
	var maxDate *int64
	err := s.db.Model(&model.Commit{}).Where("repo_id = ?", repoID).Select("MAX(commit_date)").Scan(&maxDate).Error
	if err != nil {
		return 0, err
	}
	if maxDate == nil {
		return 0, nil
	}
	return *maxDate, nil
}

func (s *CommitService) GetCommitDateRangeForRepo(repoID string) (*dto.CommitDateRange, error) {
	type result struct {
		MinDate *int64
		MaxDate *int64
	}
	var r result
	err := s.db.Model(&model.Commit{}).Where("repo_id = ?", repoID).
		Select("MIN(commit_date) as min_date, MAX(commit_date) as max_date").Scan(&r).Error
	if err != nil {
		return nil, err
	}
	if r.MinDate == nil || r.MaxDate == nil {
		return nil, nil
	}
	return &dto.CommitDateRange{Earliest: *r.MinDate, Latest: *r.MaxDate}, nil
}

type BatchInsertResult struct {
	Inserted int
	Skipped  int
}

func (s *CommitService) BatchInsertCommits(repoID string, commits []githelper.ScannedCommit) (*BatchInsertResult, error) {
	if len(commits) == 0 {
		return &BatchInsertResult{}, nil
	}

	now := model.NowMs()

	// Get unique hashes
	hashSet := make(map[string]bool)
	for _, c := range commits {
		hashSet[c.Hash] = true
	}
	hashes := make([]string, 0, len(hashSet))
	for h := range hashSet {
		hashes = append(hashes, h)
	}

	// Check existing
	var existing []model.Commit
	s.db.Where("repo_id = ? AND commit_hash IN ?", repoID, hashes).
		Select("commit_hash").Find(&existing)
	existingSet := make(map[string]bool)
	for _, e := range existing {
		existingSet[e.CommitHash] = true
	}

	// Filter out duplicates, dedup by hash (prefer first occurrence)
	byHash := make(map[string]githelper.ScannedCommit)
	for _, c := range commits {
		if existingSet[c.Hash] {
			continue
		}
		if _, ok := byHash[c.Hash]; !ok {
			byHash[c.Hash] = c
		}
	}

	if len(byHash) == 0 {
		return &BatchInsertResult{Inserted: 0, Skipped: len(commits)}, nil
	}

	// Batch create
	toInsert := make([]model.Commit, 0, len(byHash))
	for _, c := range byHash {
		branch := branchForDB(c.Branch)
		toInsert = append(toInsert, model.Commit{
			RepoID:       repoID,
			CommitHash:   c.Hash,
			AuthorName:   c.AuthorName,
			AuthorEmail:  c.AuthorEmail,
			CommitDate:   c.Date,
			Message:      c.Message,
			FilesChanged: c.FilesChanged,
			Insertions:   c.Insertions,
			Deletions:    c.Deletions,
			Branch:       branch,
			CreatedAt:    now,
		})
	}

	inserted := 0
	// Try batch insert
	err := s.db.CreateInBatches(toInsert, 100).Error
	if err != nil {
		slog.Warn("Batch insert failed, trying individual inserts", "error", err)
		for _, commit := range toInsert {
			if err := s.db.Create(&commit).Error; err != nil {
				slog.Debug("Skipped duplicate commit", "hash", commit.CommitHash)
			} else {
				inserted++
			}
		}
	} else {
		inserted = len(toInsert)
	}

	return &BatchInsertResult{Inserted: inserted, Skipped: len(commits) - inserted}, nil
}

func branchForDB(branch string) *string {
	if branch == "" || branch == "release" || branch == "master" {
		return nil
	}
	return &branch
}

func (s *CommitService) GetCommitsByDate(params struct {
	StartDate     int64
	EndDate       int64
	RepositoryIDs []string
	AuthorEmails  []string
	IsOvertime    *bool
}) (*dto.CommitsByDateResponse, error) {
	query := s.db.Model(&model.Commit{}).
		Where("commit_date >= ? AND commit_date <= ?", params.StartDate, params.EndDate)

	if len(params.RepositoryIDs) > 0 {
		query = query.Where("repo_id IN ?", params.RepositoryIDs)
	}
	if len(params.AuthorEmails) > 0 {
		query = query.Where("author_email IN ?", params.AuthorEmails)
	}

	var commits []struct {
		model.Commit
		RepoName string
	}
	query = query.Select("commits.*, repositories.name as repo_name").
		Joins("LEFT JOIN repositories ON commits.repo_id = repositories.id").
		Order("commit_date DESC")
	if err := query.Find(&commits).Error; err != nil {
		return nil, err
	}

	// Get work status config
	wsConfig, err := s.dataMetricsConfigSvc.GetWorkStatusConfig()
	if err != nil {
		wsConfig = &workstatus.DefaultConfig
	}

	// Map commits with overtime info
	commitsWithOvertime := make([]dto.CommitWithOvertime, len(commits))
	for i, c := range commits {
		isOT := workstatus.IsOvertime(c.CommitDate, wsConfig.OvertimeHour)
		var otTimes []string
		if isOT {
			otTimes = []string{workstatus.FormatTime(c.CommitDate)}
		}
		commitsWithOvertime[i] = dto.CommitWithOvertime{
			ID:                  c.ID,
			RepoID:             c.RepoID,
			RepoName:           c.RepoName,
			CommitHash:         c.CommitHash,
			AuthorName:         c.AuthorName,
			AuthorEmail:        c.AuthorEmail,
			CommitDate:         c.CommitDate,
			Message:            c.Message,
			FilesChanged:       c.FilesChanged,
			Insertions:         c.Insertions,
			Deletions:          c.Deletions,
			Branch:             c.Branch,
			CreatedAt:          c.CreatedAt,
			IsOvertime:         isOT,
			OvertimeCommitTimes: otTimes,
		}
	}

	// Group by date, dedup same commit in same day
	type dateKey struct {
		date string
		key  string // repoId:commitHash
	}
	dateMap := make(map[string][]dto.CommitWithOvertime)
	seenMap := make(map[dateKey]bool)

	for _, commit := range commitsWithOvertime {
		date := msToDateStr(commit.CommitDate)
		key := dateKey{date: date, key: commit.RepoID + ":" + commit.CommitHash}
		if seenMap[key] {
			continue
		}
		seenMap[key] = true
		dateMap[date] = append(dateMap[date], commit)
	}

	// Build result
	var result []dto.CommitsByDate
	for date, dayCommits := range dateMap {
		var overtimeCommits []dto.CommitWithOvertime
		for _, c := range dayCommits {
			if c.IsOvertime {
				overtimeCommits = append(overtimeCommits, c)
			}
		}

		// Latest overtime times (up to 5)
		var overtimeTimes []string
		sort.Slice(overtimeCommits, func(i, j int) bool {
			return overtimeCommits[i].CommitDate > overtimeCommits[j].CommitDate
		})
		for i, c := range overtimeCommits {
			if i >= 5 {
				break
			}
			if len(c.OvertimeCommitTimes) > 0 {
				overtimeTimes = append(overtimeTimes, c.OvertimeCommitTimes[0])
			}
		}

		// Sort commits by date desc
		sort.Slice(dayCommits, func(i, j int) bool {
			return dayCommits[i].CommitDate > dayCommits[j].CommitDate
		})

		hasRelease := false
		repoSet := make(map[string]bool)
		branchSet := make(map[string]bool)
		for _, c := range dayCommits {
			msgLower := strings.ToLower(c.Message)
			if strings.Contains(msgLower, "chore(release)") || strings.Contains(msgLower, "chore: release") {
				hasRelease = true
			}
			repoSet[c.RepoName] = true
			if c.Branch != nil && *c.Branch != "" {
				branchSet[*c.Branch] = true
			}
		}

		repos := make([]string, 0, len(repoSet))
		for r := range repoSet {
			repos = append(repos, r)
		}
		sort.Strings(repos)

		branches := make([]string, 0, len(branchSet))
		for b := range branchSet {
			branches = append(branches, b)
		}
		sort.Strings(branches)

		ws := workstatus.CalculateWorkStatus(len(dayCommits), len(overtimeCommits) > 0, *wsConfig)

		result = append(result, dto.CommitsByDate{
			Date:                  date,
			Commits:               dayCommits,
			TotalCommits:          len(dayCommits),
			OvertimeCount:         len(overtimeCommits),
			LatestOvertimeCommits: overtimeTimes,
			WorkStatus:            ws,
			HasRelease:            hasRelease,
			Repositories:          repos,
			Branches:              branches,
		})
	}

	// Sort by date desc
	sort.Slice(result, func(i, j int) bool {
		return result[i].Date > result[j].Date
	})

	// Filter by isOvertime if specified
	if params.IsOvertime != nil {
		filtered := make([]dto.CommitsByDate, 0)
		for _, d := range result {
			if *params.IsOvertime {
				if d.OvertimeCount > 0 {
					filtered = append(filtered, d)
				}
			} else {
				if d.OvertimeCount == 0 {
					filtered = append(filtered, d)
				}
			}
		}
		result = filtered
	}

	totalCommits := 0
	for _, d := range result {
		totalCommits += d.TotalCommits
	}

	return &dto.CommitsByDateResponse{Data: result, Total: totalCommits}, nil
}

func (s *CommitService) GetCommits(params struct {
	StartDate     int64
	EndDate       int64
	RepositoryIDs []string
	AuthorEmails  []string
	Page          int
	PageSize      int
}) (*dto.CommitsResponse, error) {
	if params.Page <= 0 {
		params.Page = 1
	}
	if params.PageSize <= 0 {
		params.PageSize = 20
	}
	offset := (params.Page - 1) * params.PageSize

	query := s.db.Model(&model.Commit{}).
		Where("commit_date >= ? AND commit_date <= ?", params.StartDate, params.EndDate)
	if len(params.RepositoryIDs) > 0 {
		query = query.Where("repo_id IN ?", params.RepositoryIDs)
	}
	if len(params.AuthorEmails) > 0 {
		query = query.Where("author_email IN ?", params.AuthorEmails)
	}

	var total int64
	query.Count(&total)

	var commits []struct {
		model.Commit
		RepoName string
	}
	query.Select("commits.*, repositories.name as repo_name").
		Joins("LEFT JOIN repositories ON commits.repo_id = repositories.id").
		Order("commit_date DESC").
		Offset(offset).Limit(params.PageSize).
		Find(&commits)

	data := make([]dto.CommitWithRepoName, len(commits))
	for i, c := range commits {
		data[i] = dto.CommitWithRepoName{
			ID:           c.ID,
			RepoID:       c.RepoID,
			RepoName:     c.RepoName,
			CommitHash:   c.CommitHash,
			AuthorName:   c.AuthorName,
			AuthorEmail:  c.AuthorEmail,
			CommitDate:   c.CommitDate,
			Message:      c.Message,
			FilesChanged: c.FilesChanged,
			Insertions:   c.Insertions,
			Deletions:    c.Deletions,
			Branch:       c.Branch,
			CreatedAt:    c.CreatedAt,
		}
	}

	return &dto.CommitsResponse{Data: data, Total: int(total), Page: params.Page, PageSize: params.PageSize}, nil
}

func (s *CommitService) GetStatistics(params struct {
	StartDate     int64
	EndDate       int64
	RepositoryIDs []string
	AuthorEmails  []string
}) (*dto.StatisticsResponse, error) {
	baseWhere := "commit_date >= ? AND commit_date <= ?"
	baseArgs := []interface{}{params.StartDate, params.EndDate}

	if len(params.RepositoryIDs) > 0 {
		placeholders := make([]string, len(params.RepositoryIDs))
		for i := range placeholders {
			placeholders[i] = "?"
		}
		baseWhere += " AND repo_id IN (" + strings.Join(placeholders, ",") + ")"
		for _, id := range params.RepositoryIDs {
			baseArgs = append(baseArgs, id)
		}
	}
	if len(params.AuthorEmails) > 0 {
		placeholders := make([]string, len(params.AuthorEmails))
		for i := range placeholders {
			placeholders[i] = "?"
		}
		baseWhere += " AND author_email IN (" + strings.Join(placeholders, ",") + ")"
		for _, email := range params.AuthorEmails {
			baseArgs = append(baseArgs, email)
		}
	}

	// Total stats
	type aggResult struct {
		Count          int64
		TotalInsertions int64
		TotalDeletions  int64
		TotalFiles      int64
	}
	var agg aggResult
	s.db.Model(&model.Commit{}).
		Where(baseWhere, baseArgs...).
		Select("COUNT(*) as count, COALESCE(SUM(insertions),0) as total_insertions, COALESCE(SUM(deletions),0) as total_deletions, COALESCE(SUM(files_changed),0) as total_files").
		Scan(&agg)

	// By repository
	type byRepoResult struct {
		RepoID       string
		Commits      int64
		Insertions   int64
		Deletions    int64
	}
	var byRepo []byRepoResult
	s.db.Model(&model.Commit{}).
		Where(baseWhere, baseArgs...).
		Select("repo_id, COUNT(*) as commits, COALESCE(SUM(insertions),0) as insertions, COALESCE(SUM(deletions),0) as deletions").
		Group("repo_id").
		Scan(&byRepo)

	// Get repo names
	repoIDs := make([]string, len(byRepo))
	for i, r := range byRepo {
		repoIDs[i] = r.RepoID
	}
	var repos []model.Repository
	s.db.Select("id, name").Where("id IN ?", repoIDs).Find(&repos)
	repoMap := make(map[string]string)
	for _, r := range repos {
		repoMap[r.ID] = r.Name
	}

	byRepository := make([]dto.ByRepository, len(byRepo))
	for i, r := range byRepo {
		name := repoMap[r.RepoID]
		if name == "" {
			name = r.RepoID
		}
		byRepository[i] = dto.ByRepository{
			RepoID:     r.RepoID,
			RepoName:   name,
			Commits:    int(r.Commits),
			Insertions: int(r.Insertions),
			Deletions:  int(r.Deletions),
		}
	}

	// By date - raw SQL
	dateWhere := baseWhere
	dateArgs := baseArgs
	dateQuery := fmt.Sprintf(`
		SELECT DATE(commit_date / 1000, 'unixepoch') as date,
			COUNT(*) as commits,
			SUM(insertions) as insertions,
			SUM(deletions) as deletions
		FROM commits
		WHERE %s
		GROUP BY date ORDER BY date DESC
	`, dateWhere)

	var byDate []dto.ByDate
	s.db.Raw(dateQuery, dateArgs...).Scan(&byDate)

	return &dto.StatisticsResponse{
		TotalCommits:    int(agg.Count),
		TotalInsertions: int(agg.TotalInsertions),
		TotalDeletions:  int(agg.TotalDeletions),
		TotalFilesChanged: int(agg.TotalFiles),
		ByRepository:    byRepository,
		ByDate:          byDate,
	}, nil
}

// ResolveScanFromDateWithGapFill extends scan window backward if DB has gaps
func ResolveScanFromDateWithGapFill(plannedRangeStartMs int64, dbLatestCommitMs int64, maxLookbackDays int) int64 {
	if maxLookbackDays <= 0 {
		maxLookbackDays = 365
	}

	now := time.Now()
	floor := time.Date(now.Year(), now.Month(), now.Day()-maxLookbackDays, 0, 0, 0, 0, time.Local)
	floorMs := floor.UnixMilli()

	if plannedRangeStartMs < floorMs {
		return floorMs
	}

	if dbLatestCommitMs == 0 || dbLatestCommitMs >= plannedRangeStartMs {
		return plannedRangeStartMs
	}

	extended := time.UnixMilli(dbLatestCommitMs).AddDate(0, 0, -2)
	extendedMs := time.Date(extended.Year(), extended.Month(), extended.Day(), 0, 0, 0, 0, time.Local).UnixMilli()

	if extendedMs < floorMs {
		return floorMs
	}
	return extendedMs
}

func msToDateStr(ms int64) string {
	return time.UnixMilli(ms).Format("2006-01-02")
}

File: python-backend/app/services/commit_service.py
"""Commit service – async SQLAlchemy port of CommitService.ts."""

from __future__ import annotations

import time
from datetime import datetime, timedelta, timezone
from typing import Any

from sqlalchemy import and_, func, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.commit import Commit
from app.models.repository import Repository
from app.services.data_metrics_config_service import DataMetricsConfigService

# ---------------------------------------------------------------------------
# Work-status calculator (mirrors workStatus.config.ts)
# ---------------------------------------------------------------------------


def calculate_work_status(
    total_commits: int,
    has_overtime: bool,
    thresholds: dict[str, int] | None = None,
) -> str:
    if thresholds is None:
        thresholds = {"relaxed": 6, "normal": 10, "busy": 15, "superCrazy": 20}

    if has_overtime:
        if total_commits >= thresholds["superCrazy"]:
            return "superCrazyOvertime"
        return "overtime"

    if total_commits < thresholds["relaxed"]:
        return "relaxed"
    elif total_commits < thresholds["normal"]:
        return "normal"
    elif total_commits < thresholds["busy"]:
        return "busy"
    elif total_commits < thresholds["superCrazy"]:
        return "crazy"
    else:
        return "crazy"


# ---------------------------------------------------------------------------
# Gap-fill resolver (mirrors resolveScanFromDateWithGapFill)
# ---------------------------------------------------------------------------


def resolve_scan_from_date_with_gap_fill(
    planned_start: datetime,
    db_latest_ms: int | None,
    max_lookback_days: int = 365,
) -> datetime:
    now = datetime.now(tz=timezone.utc)
    floor = (now - timedelta(days=max_lookback_days)).replace(
        hour=0, minute=0, second=0, microsecond=0
    )

    if planned_start < floor:
        return floor

    if db_latest_ms is None or db_latest_ms >= int(planned_start.timestamp() * 1000):
        return planned_start

    extended = datetime.fromtimestamp(db_latest_ms / 1000, tz=timezone.utc)
    extended = (extended - timedelta(days=2)).replace(
        hour=0, minute=0, second=0, microsecond=0
    )

    if extended < floor:
        return floor
    return extended


# ---------------------------------------------------------------------------
# CommitService
# ---------------------------------------------------------------------------


class CommitService:
    """Query and insert commits, compute statistics and by-date groupings."""

    # ------------------------------------------------------------------
    # max commit date
    # ------------------------------------------------------------------

    @staticmethod
    async def get_max_commit_date_ms(db: AsyncSession, repo_id: str) -> int | None:
        stmt = select(func.max(Commit.commit_date)).where(Commit.repo_id == repo_id)
        val = (await db.execute(stmt)).scalar_one_or_none()
        return int(val) if val is not None else None

    # ------------------------------------------------------------------
    # commit date range
    # ------------------------------------------------------------------

    @staticmethod
    async def get_commit_date_range(
        db: AsyncSession, repo_id: str
    ) -> dict[str, int] | None:
        stmt = select(
            func.min(Commit.commit_date).label("earliest"),
            func.max(Commit.commit_date).label("latest"),
        ).where(Commit.repo_id == repo_id)
        row = (await db.execute(stmt)).one()
        if row.earliest is None or row.latest is None:
            return None
        return {"earliest": int(row.earliest), "latest": int(row.latest)}

    # ------------------------------------------------------------------
    # batch insert
    # ------------------------------------------------------------------

    @staticmethod
    async def batch_insert_commits(
        db: AsyncSession,
        repo_id: str,
        commits: list[dict[str, Any]],
    ) -> dict[str, int]:
        if not commits:
            return {"inserted": 0, "skipped": 0}

        now = int(time.time() * 1000)

        # deduplicate by hash from input
        unique_hashes = list({c["hash"]: True for c in commits}.keys())

        # find existing hashes
        stmt = select(Commit.commit_hash).where(
            Commit.repo_id == repo_id,
            Commit.commit_hash.in_(unique_hashes),
        )
        existing_rows = (await db.execute(stmt)).scalars().all()
        existing_set = set(existing_rows)

        new_commits = [c for c in commits if c["hash"] not in existing_set]
        if not new_commits:
            return {"inserted": 0, "skipped": len(commits)}

        # keep first occurrence per hash
        by_hash: dict[str, dict] = {}
        for c in new_commits:
            if c["hash"] not in by_hash:
                by_hash[c["hash"]] = c
        to_insert = list(by_hash.values())

        def _branch_for_db(branch: str | None) -> str | None:
            if not branch or branch in ("release", "master"):
                return None
            return branch

        try:
            for commit in to_insert:
                db.add(
                    Commit(
                        repo_id=repo_id,
                        commit_hash=commit["hash"],
                        author_name=commit["authorName"],
                        author_email=commit["authorEmail"],
                        commit_date=commit["date"],
                        message=commit["message"],
                        files_changed=commit.get("filesChanged", 0),
                        insertions=commit.get("insertions", 0),
                        deletions=commit.get("deletions", 0),
                        branch=_branch_for_db(commit.get("branch")),
                        created_at=now,
                    )
                )
            await db.flush()
        except Exception:
            # fallback to individual inserts
            await db.rollback()
            inserted = 0
            for commit in to_insert:
                try:
                    db.add(
                        Commit(
                            repo_id=repo_id,
                            commit_hash=commit["hash"],
                            author_name=commit["authorName"],
                            author_email=commit["authorEmail"],
                            commit_date=commit["date"],
                            message=commit["message"],
                            files_changed=commit.get("filesChanged", 0),
                            insertions=commit.get("insertions", 0),
                            deletions=commit.get("deletions", 0),
                            branch=_branch_for_db(commit.get("branch")),
                            created_at=now,
                        )
                    )
                    await db.flush()
                    inserted += 1
                except Exception:
                    pass
            return {"inserted": inserted, "skipped": len(commits) - inserted}

        return {"inserted": len(to_insert), "skipped": len(commits) - len(to_insert)}

    # ------------------------------------------------------------------
    # helpers – overtime & time formatting
    # ------------------------------------------------------------------

    @staticmethod
    def _is_overtime(commit_date_ms: int, overtime_hour: int) -> bool:
        dt = datetime.fromtimestamp(commit_date_ms / 1000)
        return dt.hour >= overtime_hour

    @staticmethod
    def _format_time(commit_date_ms: int) -> str:
        dt = datetime.fromtimestamp(commit_date_ms / 1000)
        return dt.strftime("%H:%M")

    # ------------------------------------------------------------------
    # commits by date
    # ------------------------------------------------------------------

    @staticmethod
    async def get_commits_by_date(
        db: AsyncSession,
        start_date: int,
        end_date: int,
        repository_ids: list[str] | None = None,
        author_emails: list[str] | None = None,
        is_overtime: bool | None = None,
    ) -> dict[str, Any]:
        conditions = [
            Commit.commit_date >= start_date,
            Commit.commit_date <= end_date,
        ]
        if repository_ids:
            conditions.append(Commit.repo_id.in_(repository_ids))
        if author_emails:
            conditions.append(Commit.author_email.in_(author_emails))
        where = and_(*conditions)

        stmt = (
            select(Commit, Repository.name.label("repo_name"))
            .join(Commit.repository)
            .where(where)
            .order_by(Commit.commit_date.desc())
        )
        rows = (await db.execute(stmt)).all()

        # fetch config
        metrics_config = await DataMetricsConfigService.get_config(db)
        overtime_hour = metrics_config["overtimeHour"]
        thresholds = metrics_config["thresholds"]

        # build commit dicts
        commits_list: list[dict] = []
        for row in rows:
            commit: Commit = row[0]
            repo_name: str = row[1]
            commit_date_ms = int(commit.commit_date)
            ot = CommitService._is_overtime(commit_date_ms, overtime_hour)
            commits_list.append(
                {
                    "id": commit.id,
                    "repoId": commit.repo_id,
                    "repoName": repo_name,
                    "commitHash": commit.commit_hash,
                    "authorName": commit.author_name,
                    "authorEmail": commit.author_email,
                    "commitDate": commit_date_ms,
                    "message": commit.message,
                    "filesChanged": commit.files_changed,
                    "insertions": commit.insertions,
                    "deletions": commit.deletions,
                    "branch": commit.branch,
                    "createdAt": int(commit.created_at),
                    "isOvertime": ot,
                    "overtimeCommitTimes": [CommitService._format_time(commit_date_ms)] if ot else None,
                }
            )

        # group by date, dedup by repoId:commitHash
        date_map: dict[str, list[dict]] = {}
        seen_in_date: dict[str, dict[str, dict]] = {}

        for c in commits_list:
            dt = datetime.fromtimestamp(c["commitDate"] / 1000)
            date_str = dt.strftime("%Y-%m-%d")
            key = f"{c['repoId']}:{c['commitHash']}"

            if date_str not in seen_in_date:
                seen_in_date[date_str] = {}

            dm = seen_in_date[date_str]
            if key in dm:
                existing = dm[key]
                if not existing.get("branch") and c.get("branch"):
                    existing["branch"] = c["branch"]
                continue

            dm[key] = c
            date_map.setdefault(date_str, []).append(c)

        # build per-day summaries
        data: list[dict] = []
        for date_str, day_commits in date_map.items():
            overtime_commits = [c for c in day_commits if c["isOvertime"]]
            overtime_times = sorted(
                [
                    {
                        "time": CommitService._format_time(c["commitDate"]),
                        "ts": c["commitDate"],
                    }
                    for c in overtime_commits
                ],
                key=lambda x: x["ts"],
                reverse=True,
            )[:5]
            overtime_times_str = [o["time"] for o in overtime_times]

            sorted_commits = sorted(day_commits, key=lambda x: x["commitDate"], reverse=True)

            has_release = any(
                "chore(release)" in c["message"].lower() or "chore: release" in c["message"].lower()
                for c in day_commits
            )

            work_status = calculate_work_status(
                len(day_commits),
                len(overtime_commits) > 0,
                thresholds,
            )

            repositories = sorted({c["repoName"] for c in day_commits})
            branches_set: set[str] = set()
            for c in day_commits:
                if c.get("branch"):
                    branches_set.add(c["branch"])
            branches = sorted(branches_set)

            data.append(
                {
                    "date": date_str,
                    "commits": sorted_commits,
                    "totalCommits": len(day_commits),
                    "overtimeCount": len(overtime_commits),
                    "latestOvertimeCommits": overtime_times_str,
                    "workStatus": work_status,
                    "hasRelease": has_release,
                    "repositories": repositories,
                    "branches": branches,
                }
            )

        # sort descending
        data.sort(key=lambda d: d["date"], reverse=True)

        # filter by overtime flag
        if is_overtime is not None:
            data = [
                d for d in data if (d["overtimeCount"] > 0) == is_overtime
            ]

        total_commits = sum(d["totalCommits"] for d in data)
        return {"data": data, "total": total_commits}

    # ------------------------------------------------------------------
    # commits (paginated)
    # ------------------------------------------------------------------

    @staticmethod
    async def get_commits(
        db: AsyncSession,
        start_date: int,
        end_date: int,
        repository_ids: list[str] | None = None,
        author_emails: list[str] | None = None,
        page: int = 1,
        page_size: int = 20,
    ) -> dict[str, Any]:
        conditions = [
            Commit.commit_date >= start_date,
            Commit.commit_date <= end_date,
        ]
        if repository_ids:
            conditions.append(Commit.repo_id.in_(repository_ids))
        if author_emails:
            conditions.append(Commit.author_email.in_(author_emails))
        where = and_(*conditions)

        # total
        count_q = select(func.count()).select_from(Commit).where(where)
        total = (await db.execute(count_q)).scalar_one()

        # data
        data_q = (
            select(Commit, Repository.name.label("repo_name"))
            .join(Commit.repository)
            .where(where)
            .order_by(Commit.commit_date.desc())
            .offset((page - 1) * page_size)
            .limit(page_size)
        )
        rows = (await db.execute(data_q)).all()

        data = [
            {
                "id": row.Commit.id,
                "repoId": row.Commit.repo_id,
                "repoName": row.repo_name,
                "commitHash": row.Commit.commit_hash,
                "authorName": row.Commit.author_name,
                "authorEmail": row.Commit.author_email,
                "commitDate": int(row.Commit.commit_date),
                "message": row.Commit.message,
                "filesChanged": row.Commit.files_changed,
                "insertions": row.Commit.insertions,
                "deletions": row.Commit.deletions,
                "branch": row.Commit.branch,
                "createdAt": int(row.Commit.created_at),
            }
            for row in rows
        ]

        return {
            "data": data,
            "total": total,
            "page": page,
            "pageSize": page_size,
        }

    # ------------------------------------------------------------------
    # statistics
    # ------------------------------------------------------------------

    @staticmethod
    async def get_statistics(
        db: AsyncSession,
        start_date: int,
        end_date: int,
        repository_ids: list[str] | None = None,
        author_emails: list[str] | None = None,
    ) -> dict[str, Any]:
        conditions = [
            Commit.commit_date >= start_date,
            Commit.commit_date <= end_date,
        ]
        if repository_ids:
            conditions.append(Commit.repo_id.in_(repository_ids))
        if author_emails:
            conditions.append(Commit.author_email.in_(author_emails))
        where = and_(*conditions)

        # aggregates
        agg_q = select(
            func.count().label("cnt"),
            func.coalesce(func.sum(Commit.insertions), 0).label("total_ins"),
            func.coalesce(func.sum(Commit.deletions), 0).label("total_del"),
            func.coalesce(func.sum(Commit.files_changed), 0).label("total_files"),
        ).where(where)
        agg = (await db.execute(agg_q)).one()

        # by repository
        by_repo_q = (
            select(
                Commit.repo_id,
                func.count().label("cnt"),
                func.coalesce(func.sum(Commit.insertions), 0).label("ins"),
                func.coalesce(func.sum(Commit.deletions), 0).label("dels"),
            )
            .where(where)
            .group_by(Commit.repo_id)
        )
        by_repo_rows = (await db.execute(by_repo_q)).all()
        repo_ids = [r.repo_id for r in by_repo_rows]

        repo_names: dict[str, str] = {}
        if repo_ids:
            rn_q = select(Repository.id, Repository.name).where(Repository.id.in_(repo_ids))
            for rr in (await db.execute(rn_q)).all():
                repo_names[rr.id] = rr.name

        by_repository = [
            {
                "repoId": r.repo_id,
                "repoName": repo_names.get(r.repo_id, r.repo_id),
                "commits": r.cnt,
                "insertions": int(r.ins),
                "deletions": int(r.dels),
            }
            for r in by_repo_rows
        ]

        # by date – raw SQL for SQLite date formatting
        params: dict[str, Any] = {"start_date": start_date, "end_date": end_date}
        by_date_sql = (
            "SELECT "
            "  DATE(commit_date / 1000, 'unixepoch') as date, "
            "  COUNT(*) as commits, "
            "  SUM(insertions) as insertions, "
            "  SUM(deletions) as deletions "
            "FROM commits "
            "WHERE commit_date >= :start_date AND commit_date <= :end_date"
        )
        if repository_ids:
            placeholders = ",".join([f":repo_{i}" for i in range(len(repository_ids))])
            by_date_sql += f" AND repo_id IN ({placeholders})"
            for i, rid in enumerate(repository_ids):
                params[f"repo_{i}"] = rid
        if author_emails:
            placeholders = ",".join([f":author_{i}" for i in range(len(author_emails))])
            by_date_sql += f" AND author_email IN ({placeholders})"
            for i, email in enumerate(author_emails):
                params[f"author_{i}"] = email
        by_date_sql += " GROUP BY date ORDER BY date DESC"

        by_date_rows = (await db.execute(text(by_date_sql), params)).mappings().all()
        by_date = [
            {
                "date": row["date"],
                "commits": int(row["commits"]),
                "insertions": int(row["insertions"]) if row["insertions"] is not None else 0,
                "deletions": int(row["deletions"]) if row["deletions"] is not None else 0,
            }
            for row in by_date_rows
        ]

        return {
            "totalCommits": agg.cnt,
            "totalInsertions": int(agg.total_ins),
            "totalDeletions": int(agg.total_del),
            "totalFilesChanged": int(agg.total_files),
            "byRepository": by_repository,
            "byDate": by_date,
        }
